20240808-custom-limit-email-notifications.md

```diff
@@ -0,0 +1,251 @@
+    * enqueue an event to the project notifications queue indicating that the notifications flag should be reset for the relevant threshold 
+
+More specifically, 
+
+When limit validation occurs for an upload:
+* if limit emails are disabled globally, or disabled for the project, skip
+* for segment and custom storage limits:
+  * for each limit threshold (starting at highest threshold):
+    * if this upload causes usage to meet or cross threshold:
+      - enqueue `(projectID, thresholdEventKind)` to `project_events`
+      - skip checking lower thresholds
+	* if usage is below the threshold, but the project is flagged indicating that the threshold was exceeded:
+      - enqueue `(projectID, thresholdEventKind)` to `project_events` ("threshold event kind" would indicate to reset the flag with no notification)
+
+When limit validation occurs for a download:
+* if limit emails are disabled globally, or disabled for the project, skip
+* for each limit threshold (starting at highest threshold):
+  * if this download causes egress to meet or cross custom egress threshold: 
+    - enqueue `(projectID, thresholdEventKind)` to `project_events`
+    - skip checking lower thresholds
+  * if egress usage is below the custom threshold, but the project is flagged indicating that the threshold was exceeded:
+    - enqueue `(projectID, thresholdEventKind)` to `project_events` ("threshold event kind" would indicate to reset the flag with no notification)
+
+While the intended behavior for event detection is straightforward, implementation may be tricky. As an example, metainfo validates whether upload limits would be exceeded for a new object in the metainfo package [here](https://github.com/storj/storj/blob/8d9fab8825e92c4c77252e3891530e5d04c430eb/satellite/metainfo/validation.go#L574) - calling a function `ExceedsUploadLimits` in accounting/projectusage.go [here](https://github.com/storj/storj/blob/b552501c39352ba80487b6991975e8d13f549410/satellite/accounting/projectusage.go#L125). The project usage service retrieves the current storage and segment usage from a [redis cache](https://github.com/storj/storj/blob/b552501c39352ba80487b6991975e8d13f549410/satellite/accounting/live/redis.go#L66) in accounting/live/redis.go.
+
+Our current live accounting project usage behavior is designed to allow for very quickly getting and updating a handful of values per project (specific limits and corresponding usage) with minimal DB calls.
+
+We need to have additional information during limit validation to properly queue events: the `notification_flags` flags from the `projects` table. This is because we not only need to check (1) is current usage above or below the threshold? We also need to check (2) are notification emails for this threshold enabled? And (3) has a notification been triggered for this threshold before?
```

Comment:
Could we off load this responsibility to the email chore? So metainfo will only queue events, then the chore checks if the event should be processed based on whether the threshold notification is enabled and if notification has been already sent.

Reply:
The email reminders chore only runs every 24 hours. So if we wanted to add it there, we would still need to add a separate loop that runs more frequently.
Anyway, the section titled "Queue processing in core" includes the text "Create a chore on the satellite (or update existing console emailreminders chore)", so this option is accounted for in the existing design.